<div style="border:solid green 2px; padding: 20px">

<b>Привет, Катя! Поздравляю с выполнением очередного проекта!</b>

Меня зовут Петр Раздобреев и я буду проверять твою работу. Попробую помочь сделать тебе её еще лучше! В Практикуме принято общени на ты, но если это дискомфортно, то сообщи об этом)

Ниже ты найдешь мои комментарии - пожалуйста, не перемещай, не изменяй и не удаляй их. Я буду использовать цветовую разметку:

<div class="alert alert-danger">
<b>Комментарий ревьюера:</b>☁️Так выделены самые важные замечания. Без их отработки проект не будет принят. </div>

<div class="alert alert-warning">
<b>Комментарий ревьюера:</b>⛅️ Так выделены небольшие замечания.Такие замечания в проекте допустимы, это пространство для улучшений в будущем.
</div>

<div class="alert alert-success">
<b>Комментарий ревьюера:</b>☀️Так я выделяю успешные идеи.</div>

Давай работать над проектом в диалоге: если ты что-то меняешь в проекте или отвечаешь на мои комменатри — пиши об этом. Мне будет легче отследить изменения, если ты выделишь свои комментарии:
<div class="alert alert-info"> <b>Комментарий студента:</b> Например, вот так.</div>

Всё это поможет выполнить повторную проверку твоего проекта оперативнее. 
    

In [48]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine
#!pip install SQLAlchemy==1.4
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

In [49]:
query_books = ''' SELECT * FROM books ''' 
query_publishers = ''' SELECT * FROM publishers ''' 
query_authors = ''' SELECT * FROM authors ''' 
query_ratings = ''' SELECT * FROM ratings ''' 
query_reviews = ''' SELECT * FROM reviews ''' 

books = pd.io.sql.read_sql(query_books, con=engine) 
publishers= pd.io.sql.read_sql(query_publishers, con=engine)
authors=pd.io.sql.read_sql(query_authors, con=engine)
ratings=pd.io.sql.read_sql(query_ratings, con=engine)
reviews= pd.io.sql.read_sql(query_reviews, con=engine)

print("Таблица с книгами")
print(books.head())
print("Таблица издательств")
print(publishers.head())
print("Таблица авторов")
print(authors.head())
print("Таблица рейтинга книг")
print(ratings.head())
print("Таблица отзывов")
print(reviews.head())

Таблица с книгами
   book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268  
Таблица издательств
   publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4          

<div class="alert alert-warning">
<b>Комментарий ревьюера:</b>⛅️ 

По другим датафреймам бы тоже вывод хотя бы добавить 
</div>

Готово :)

### задание. 
Посчитайте, сколько книг вышло после 1 января 2000 года

In [50]:
query_1 = ''' SELECT count(book_id) as "Вышло книг с после 1 янв.2000г" FROM books  
              where CAST (publication_date as date) > '2000-01-01' ''' 

books_cnt = pd.io.sql.read_sql(query_1, con=engine) 

print(books_cnt)

   Вышло книг с после 1 янв.2000г
0                             819


<div class="alert alert-success">
<b>Комментарий ревьюера:</b>☀️
    
Абсолютно верно

После 1 января 2000 года было выпущено 819 книг.

### Задание.
Для каждой книги посчитайте количество обзоров и среднюю оценку

In [51]:
query_2 = ''' SELECT Main.title as "Название книнги", RW.reviews as "Обзоров на книгу", R.rate as "Средний рейтинг"
              FROM books Main 
              left join (select book_id, count (review_id) as reviews from reviews GROUP by book_id) as RW on Main.book_id = RW.book_id
              left join (select book_id, AVG(rating) as rate from ratings GROUP by book_id) as R on Main.book_id = R.book_id
              order by 3 desc
          ''' 

books_rate = pd.io.sql.read_sql(query_2, con=engine) 

books_rate

Название книнги  Обзоров на книгу  \
0                 Pop Goes the Weasel (Alex Cross  #5)               2.0   
1    The Ghost Map: The Story of London's Most Terr...               2.0   
2    In the Hand of the Goddess (Song of the Liones...               2.0   
3                             Tai-Pan (Asian Saga  #2)               2.0   
4    How to Be a Domestic Goddess: Baking and the A...               1.0   
..                                                 ...               ...   
995  The World Is Flat: A Brief History of the Twen...               3.0   
996                                              Junky               2.0   
997                  His Excellency: George Washington               2.0   
998                                      Drowning Ruth               3.0   
999                               Harvesting the Heart               2.0   

     Средний рейтинг  
0               5.00  
1               5.00  
2               5.00  
3               5.00  
4               5.00  
..               ...  
995             2.25  
996             2.00  
997             2.00  
998             2.00  
999             1.50  

[1000 rows x 3 columns]

<div class="alert alert-success">
<b>Комментарий ревьюера:</b>☀️
    
Рекомендовал бы сортировать по числу ревью, чтобы дополнительно проверять, что не получилось каких-то экстра больших значений </div>

In [52]:
a = books_rate['Обзоров на книгу'].describe()
a

count    994.000000
mean       2.809859
std        1.055873
min        1.000000
25%        2.000000
50%        3.000000
75%        3.000000
max        7.000000
Name: Обзоров на книгу, dtype: float64

In [53]:
b = books_rate['Средний рейтинг'].describe()
b

count    1000.000000
mean        3.898973
std         0.562376
min         1.500000
25%         3.500000
50%         4.000000
75%         4.333333
max         5.000000
Name: Средний рейтинг, dtype: float64

- КОЛИЧЕСТВО ОБЗОРОВ. В среднем, на одну книгу составлено 2,8 обзора. Максимальное число обзоров на книгу - 7. Минимальное - 1.
- СРЕДНЯЯ ОЦЕНКА КНИГИ. Средний рейтинг книги - 3,89. Минимальный рейтинг - 1,5. Максимальный - 5. 

### Задание. 
Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры

In [54]:
query_3 = ''' with Temp as 
                (SELECT count(FAT.book_id) as cnt, P.publisher
                from (select * FROM books where num_pages > 50) as FAT
                join publishers P on FAT.publisher_id=P.publisher_id
                group by publisher)
              select cnt as "Количество книг", 
              publisher as "Издательство"
              from Temp
              where cnt = (select MAX(cnt) from Temp)              
          ''' 

top_pub = pd.io.sql.read_sql(query_3, con=engine) 

top_pub

Количество книг   Издательство
0               42  Penguin Books

Издательство, выпустившее наибольшее число книг толще 50 страниц -  "Penguin Books". На их счету выпуск 42 таких книг.

<div class="alert alert-success">
<b>Комментарий ревьюера:</b>☀️
    
Издательство также определено верно!</div>

### Задание.
Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками

In [55]:
query_4 = ''' with RATE as (select count(rating_id) as rate_cnt, book_id from ratings group by book_id),
              RATE_AVG as (select book_id, AVG(rating) as AVG_rate from ratings GROUP by book_id)
                select author as "Автор", AVG(AVG_rate) as "Средняя оценка" 
                from books 
                join authors on books.author_id = authors.author_id
                join RATE on books.book_id = RATE.book_id
                join RATE_AVG on books.book_id = RATE_AVG.book_id
                where rate_cnt > 50
                GROUP by author
                order by 2 desc
                limit 3
          ''' 

top_author = pd.io.sql.read_sql(query_4, con=engine) 

top_author

Автор  Средняя оценка
0         J.K. Rowling/Mary GrandPré        4.283844
1  Markus Zusak/Cao Xuân Việt Khương        4.264151
2                     J.R.R. Tolkien        4.258446

<div class="alert alert-danger">
<b>Комментарий ревьюера:</b>☁️

Тут проблема в том, что ожидался средний рейтинг автора по всем его книгам, а вот даже в топе нескольки книг мамы Гарри Поттера
</div>

<div class="alert alert-success">
<b>Комментарий ревьюера v2:</b>☀️
   
Отлично, теперь верно. Ниже обратная связь по результатам взаимодействия. 

Благодарю за совместную работу и хочу поздравить, ты большая молодец, что прошла такой путь и освоила программу, надеюсь у тебя сбудутся все намеченные планы.

У Практикума развивается сообщество выпускников, так что рекомендую про него не забывать, в случае необходимости.


Рекомендовал бы для первичного анализа использовать `pandas profiling`, удобная библиотека для автоматизации как раз первичного анализа. Возможно набор ниже для вдохновения, покажется интересным для тебя и ты реализуешь. Если планируешь собеседоваться, то рекомендую подтянуть `SQL`, задачки на собеседованиях обычно на него любят давать. Желаю удачи!


Оформление презентаций дело субъективное и зачастую в разных компаниях имеются свои требования, но вот видео от Яндекса по созданию презентаций https://www.youtube.com/watch?v=S0r0fMJa9eA, а ещё на платформе практикума открылся бесплатный курс по созданию презентаций, рекомендую.

Для более глубокого изучения tableau, в случае необходимости рекомендую материал https://tableau.pro/m01, этот курс делает классная команда, у них бывают и другие интересные мини курсы. Но в последнее время российские компании уходят от tableau или уже ушли, поэтому возможно придётся познакомиться с другими инструментами.

Задачи на статистику или вокруг неё часто возникают на собеседовании, для укрепления этого фундамента рекомендую статью для выбора критерия проверки гипотез https://lit-review.ru/biostatistika/vybor-statisticheskogo-kriteriya/, https://statpsy.ru/ - информация по критериям проверки гипотез.

Визуализации данных для аналитиков очень важны, при помощи них легче доносить желаемые мысли, а для того, чтобы лучше отобразить, нужно понимать, какие вообще существуют визуализации, делюсь с тобой своей подборкой для разных случае жизни https://www.notion.so/d9392e1cb71f48cfb1fbf7f557a2b7e7 

SQL важный навык, который нужно поддерживать для этого существует много тренажёров:
* Статья на хабр с топ-тренажёрами: https://techrocks.ru/2019/11/22/7-sites-to-perfect-sql-query-writing-skills/
* задачки SQL - https://habr.com/ru/company/dcmiran/blog/500360/
* https://sqlzoo.net/

Согласно результатам проведённого анализа, в топ-3 авторов с самой высокой средней оценкой книг входят не только известные всем создатели серий книг о "мальчике, который выжил" и удивительных хоббитах, но и австралийский писатель Markus Zusak. Но все же возглавляет рейтинг J.K. Rowling/Mary GrandPré.

### Задание.
Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок

В схеме БД нет точной информации о связях таблиц ratings и reviews по полю username, но можно проверить по количеству уникальных пользователей в обеих таблицах, а также сопоставить их имена:

In [56]:
query_pre_5 = ''' select * from (select distinct(username) from ratings) R
                  join
                  (select distinct(username) from reviews) RW 
                  on R.username=RW.username
              '''

users = pd.io.sql.read_sql(query_pre_5, con=engine) 

users 

username        username
0           jnelson         jnelson
1         zjohnston       zjohnston
2      lewisdesiree    lewisdesiree
3     patrickhudson   patrickhudson
4             ulowe           ulowe
..              ...             ...
155         dmiller         dmiller
156  shermannatalie  shermannatalie
157  charlesraymond  charlesraymond
158     isaiahreyes     isaiahreyes
159       masonsara       masonsara

[160 rows x 2 columns]

По результатам сравнения мжно считать, что пользователи в обеих таблицах идентичны.

In [57]:
query_5 = ''' with RATE as (select count(rating_id) as rate_cnt, username from ratings group by username),
              REV as (select count(review_id) as rev_cnt, username from reviews group by username)
                select AVG(rev_cnt) as "Среднее кол-во обзоров"
                from RATE join REV on REV.username = RATE.username
                where rate_cnt > 50
          ''' 

avg_reviews = pd.io.sql.read_sql(query_5, con=engine) 

avg_reviews 

Среднее кол-во обзоров
0               24.333333

<div class="alert alert-success">
<b>Комментарий ревьюера:</b>☀️
    
И это правильный ответ, чувствую себя Дибровым. 
    

Среднее количество обзоров от пользователей, которые поставили больше 50 оценок - 24,33.